aiida/tools/graph/age_entities.py

```diff
@@ -0,0 +1,508 @@
+        # Done with checks, saving to attributes:
+        self._aiida_cls_to = aiida_cls_to
+        self._aiida_cls_from = aiida_cls_from
+        # The _set is the set where keys are set:
+        self.keyset = set()
+
+        # I need to get the identifiers for the edge. For now, these should be hardcoded
+        if aiida_cls_from is Node:
+            if aiida_cls_to is Node:
+                # This is a Node-Node connection, therefore I will get information on the links
+                self._edge_identifiers = (('edge', 'input_id'), ('edge', 'output_id'), ('edge', 'type'),
+                                          ('edge', 'label'))
+                self._edge_namedtuple = LinkQuadruple
+            elif aiida_cls_to is Group:
+                self._edge_identifiers = (('nodes', 'id'), ('groups', 'id'))
+                self._edge_namedtuple = GroupNodeEdge
+            else:
+                raise TypeError(
+                    'Unexpted types aiida_cls_from={} and aiida_cls_to={}'.format(aiida_cls_from, aiida_cls_to)
+                )
+        else:
+            raise TypeError('Unexpted types aiida_cls_from={} and aiida_cls_to={}'.format(aiida_cls_from, aiida_cls_to))
+
+    def _check_self_and_other(self, other):
+        """
+        Utility function. When called, will check whether self and other instance
+        are compatible. Same aiida classes, same identifiers, etc...
+        """
+        if not isinstance(other, DirectedEdgeSet):
+            raise TypeError('Other class is not an instance of AiidaEntitySet')
+        if self.aiida_cls_to != other.aiida_cls_to:
+            raise TypeError('The two instances do not have the same aiida type!')
+        if self.aiida_cls_from != other.aiida_cls_from:
+            raise TypeError('The two instances do not have the same aiida type!')
+        if self._edge_namedtuple != other.get_edge_namedtuple():
+            raise ValueError('The two instances do not have the same identifiers!')
+        return True
+
+    def _check_input_for_set(self, input_for_set):
+        """
+        When giving me something to the set, this utility function can be used
+        to do the right thing.
+        """
+        if isinstance(input_for_set, tuple):
+            if len(input_for_set) != len(self._edge_identifiers):
+                raise ValueError(
+                    'The tuple you passed has not the right length {} != {}'.format(
+                        len(input_for_set), len(self._edge_identifiers)
+                    )
+                )
+            return input_for_set
+
+        raise TypeError('{} is not a valid input\n' 'It has to be a tuple'.format(input_for_set))
+
+    def get_template(self):
+        """
+        Create a new instance with the same defining attributes.
+        """
+        new = DirectedEdgeSet(aiida_cls_to=self.aiida_cls_to, aiida_cls_from=self.aiida_cls_from)
+        return new
+
+    def copy(self):
```

Comment:
Does it make sense to merge `copy` and `get_template` and use `keyset=False` as an argument to also copy over the keyset or not

Reply:
Paging @zooks97 and @lekah : I know this was the way in which this used to work originally and you had your reasons for changing it, correct?

Reply:
I think it's better to have a distinction between the two operations:

- `copy` gives you an identical copy, with all data, of the AiidaEntitySet / Basket
- `get_template` gives you a new instance of the AiidaEntitySet / Basket that is parameterized in the same way (similar to having the same `dtype` in a numpy `Array`) but is empty